src/bag/simulation/cache.py

```diff
@@ -330,7 +359,132 @@ async def _create_dut(self, impl_cell: str,
+        is_cached = False
+
+        # In an interface, the best "uniqueness" is based on a cell's library and name
+        # We will use this same strategy.
+        hash_id = hash(f'{lib_name}__{impl_cell}')
+
+        dir_list = self._cache.get(hash_id, None)
+        if dir_list is None:
+            dir_list = []
+            self._cache[hash_id] = dir_list
+            dir_path = self._generate_cell(impl_cell, cdl_netlist, gds_file)
+            dir_list.append(dir_path.name)
+            write_yaml(self._info_file, self._info_specs)
+        else:
+            dir_path = None
+            dirs_to_remove = []
+            for dir_name in dir_list:
+                cur_dir = self._root_dir / dir_name
+                cur_netlist = cur_dir / 'netlist.cdl'
+                if not cur_netlist.exists():  # Cached netlist is missing, should ignore and remove from the cache
+                    dirs_to_remove.append(dir_name)
+                    continue
+                netlist_cached = await self.cdl_check_cache(impl_cell, cdl_netlist, cur_dir)
+                if netlist_cached:
+                    is_cached = await self.gds_check_cache(gds_file, cur_dir)
+                    if is_cached:
+                        self.log('Found existing design, reusing DUT netlist.')
+                        dir_path = cur_dir
+                        break
+
+            if dirs_to_remove:
+                for dir_name in dirs_to_remove:
+                    dir_list.remove(dir_name)
+                if dir_list:  # Remove entry from cache
+                    del self._cache[hash_id]
+                else:
+                    self._cache[hash_id] = dir_list
+                write_yaml(self._info_file, self._info_specs)
+
+            if dir_path is None:
+                dir_path = self._generate_cell(impl_cell, cdl_netlist, gds_file)
+                dir_list.append(dir_path.name)
+                write_yaml(self._info_file, self._info_specs)
+
+        shutil.rmtree(tmp_dir)  # Remove temporary directory
+        if em:
+            ans = Path(cdl_netlist)
+            extract_info = dir_path / 'layout.gds'
+        else:
+            if extract or (extract is None and self._extract):
+                ans = dir_path / 'rcx.sp'
+                if not ans.exists() or self._force_extract:
+                    extract_info = dir_path
+                else:
+                    extract_info = None
+            else:
+                extract_info = None
+                ans = dir_path / 'netlist.cdl'
+                """
+                CDL netlists can be simulated. StrmOut puts PDK library cells in the netlist. This
+                conflicts with BAG's netlisting, which puts the library cells in `sim.scs`; the
+                double placement causes a conflict at Spectre runtime. One of these has to be
+                resolved before simulating with CDL.
+                """
+                raise RuntimeError("Not implemented yet. See developer")
```

Comment:
Are you sure this is because of the PDK library cells or the BAG_prim cells? I think it's the latter. Including a file twice is usually not an issue. If it is indeed the BAG_prim definitions, I think the easiest solution is to change the C++ netlisting to include a file that has all BAG_prim definitions (stored in the netlist_setup directory), rather than writing the definitions at the top of the netlist. Do this for both CDL and Spectre.

Reply:
Not sure at the moment. I will mark an issue to resolve.

Reply:
For a bit more context, `BAG_prim` cell definitions used to exist in both `.cdl` and `.scs` netlists in the original implementation. But when harnesses were implemented, it became an issue for schematic harnesses. The `BAG_prim` cells would get defined in both the DUT netlist as well as harness netlist(s), causing a Spectre error. So I modified pybag / cbag to avoid putting `BAG_prim` cell definitions in the `.scs` netlists, and instead place it only once in the top level simulation netlist `sim.scs`. However I kept the `BAG_prim` cell definitions in the `.cdl` netlists since that is necessary for LVS, and we weren't using CDL for simulations previously.
Hence, removing `BAG_prim` cell definitions from `sim.scs` is not a valid solution, because of harnesses. Removing those from `.cdl` is also not a valid solution, because of LVS. So 2 options are:
1. Put some unique prefixes / suffixes on the `BAG_prim` cell names when they are being copied into the `.cdl` netlist. This can be quite difficult to implement.
2. (Better option): `include` a file with the `BAG_prim` cell definitions instead of writing it in every file. I think this can solve both the problems, but we need to check.